api/integrations/slack/views.py

```diff
@@ -0,0 +1,92 @@
+import uuid
+
+from django.conf import settings
+from django.core.exceptions import ObjectDoesNotExist
+from django.shortcuts import redirect
+from rest_framework import status
+from rest_framework.decorators import action
+from rest_framework.response import Response
+from slack_sdk.oauth import AuthorizeUrlGenerator
+
+from integrations.common.views import IntegrationCommonViewSet
+from integrations.slack.models import SlackConfiguration, SlackEnvironment
+from integrations.slack.serializers import (
+    SlackChannelListSerializer,
+    SlackEnvironmentSerializer,
+)
+from integrations.slack.slack import get_bot_token, get_channels_data
+
+
+class SlackConfigurationViewSet(IntegrationCommonViewSet):
+    serializer_class = SlackEnvironmentSerializer
+    pagination_class = None  # set here to ensure documentation is correct
+    model_class = SlackEnvironment
+
+    @action(
+        detail=False,
+        methods=["GET"],
+        url_path="channels",
+        serializer_class=SlackChannelListSerializer,
+    )
+    def get_channels(self, request, *args, **kwargs):
+        env = self.get_environment_from_request()
+        try:
+            config = SlackConfiguration.objects.get(project=env.project)
+        except ObjectDoesNotExist:
+            return Response(
+                "Slack api token not found. Please generate the token using oauth",
+                status.HTTP_400_BAD_REQUEST,
+            )
+        slack_token = config.api_token
+        serializer = self.get_serializer(data=get_channels_data(slack_token), many=True)
+        serializer.is_valid()
+        return Response(serializer.data)
+
+    def _get_slack_callback_url(self, **kwargs):
+        return self.reverse_action("slack-oauth-callback", kwargs=kwargs)
+
+    @action(detail=False, methods=["GET"], url_path="callback", permission_classes=[])
+    def slack_oauth_callback(self, request, *args, **kwargs):
+        code = request.GET.get("code")
+        if not code:
+            return Response(
+                "code not found in query params",
+                status.HTTP_400_BAD_REQUEST,
+            )
+
+        env = self.get_environment_from_request()
+        validate_state(request.GET.get("state"), request)
+        bot_token = get_bot_token(code, self._get_slack_callback_url(**kwargs))
```

Comment:
this will be kwargs captured from url. in this case something like: `{'environment_api_key': 'xyz''}`

Reply:
Gotcha ok. I don't really like the fact that we're blindly passing in the kwargs like this though. I think it would be more readable if we add in an extract line to extract the environment_api_key into a variable from the kwargs.